AX: isolated tree nodes should store frequent bool attributes in a bitfield

minorninth opened the pull request:
#### 48e4bde9c8cf65a3e06b932522e17a97f6acc0c1
<pre>
AX: isolated tree nodes should store frequent bool attributes in a bitfield
<a href="https://bugs.webkit.org/show_bug.cgi?id=275113">https://bugs.webkit.org/show_bug.cgi?id=275113</a>
<a href="https://rdar.apple.com/129223343">rdar://129223343</a>

Reviewed by Andres Gonzalez.

A lot of memory is wasted in AX isolated tree nodes storing boolean
attributes in an expensive HashMap. By storing some of the most common
ones in a bitfield we can save memory. The most rare ones are still a
good fit for the HashMap.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::AXIsolatedObject):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::boolAttributeValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
(WebCore::AXIsolatedObject::setPropertyFlag):
(WebCore::AXIsolatedObject::hasPropertyFlag const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/279938@main">https://commits.webkit.org/279938@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c0f31df434fade1959d35e206f8863eecfd0b54

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54983 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34440 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7576 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58261 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5714 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57283 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41990 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5746 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44530 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3888 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47637 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25656 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29312 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4981 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3855 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5202 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59852 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30250 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/5361 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51950 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31380 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47713 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51392 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed, /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12086 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31171 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -446,6 +446,86 @@ void AXIsolatedObject::setObjectVectorProperty(AXPropertyName propertyName, cons
 void AXIsolatedObject::setProperty(AXPropertyName propertyName, AXPropertyValueVariant&& value)
 {
+    switch (propertyName) {
+    case AXPropertyName::CanSetFocusAttribute:
```
Maybe this is an unnecessary micro-optimization. But I wonder if we should gate this switch statement with tons of branches on whether `value` is actually a `bool`? That way every non-bool property doesn't do all of this checking.

minorninth replied:
That's a good thought!

OK, check it out now. Slightly different idea, let's just reuse the same enum but put the bool flags at the top. That's easier to maintain and doesn't require any switch statements at all.

twilco reviewed (approved):
Looks good to me overall, this is a big win! Thanks for doing it. Hoping @AndresGonzalezApple can have a chance to review it too.

AndresGonzalezApple commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
```diff
@@ -552,6 +552,8 @@ class AXIsolatedObject final : public AXCoreObject {
     Vector<RefPtr<AXCoreObject>> m_children;
     AXPropertyMap m_propertyMap;
+    // A bitfield for some of the most common bool properties
+    uint32_t m_propertyFlags;
```
This is a clever and useful optimization. I would suggest for maintainability and clarity sake to consider the following implementation approach:
1. Leave the AXPropertyName enum as is.
2. Add a new enum with powers of 2 called something like CommonBoolPropertyBit. This can be a private enum in AXIsolatedObject.
3. A simple helper function to return a CommonBoolPropertyBit given a PropertyName or 0 if not a common bool. This would be essentially a switch statement.
4. Make m_propertyFlags and WTF::OptionSet<CommonBoolPropertyBit>.
5. Use WTF::OptionSet methods for manipulation of the bits.
I think this would make it more clear and robust down the road, and still be a good optimization.

minorninth replied:
Yeah, my original version was like that. After Tyler's feedback I decided to try to optimize it more but maybe it's less maintainable.

I put it back to the original version that uses OptionSet and switch statements. Let me know what you think.

AndresGonzalezApple commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -505,6 +587,29 @@ void AXIsolatedObject::setProperty(AXPropertyName propertyName, AXPropertyValueV
+{
+    return m_propertyFlags.contains(flag);
+}
+
```
Maybe all these one line methods should be inline in the header file below the class declaration. Maybe that's now done by the compiler automatically.

minorninth replied:
Sounds good. Inlined.

minorninth committed:
AX: isolated tree nodes should store frequent bool attributes in a bitfield
https://bugs.webkit.org/show_bug.cgi?id=275113
rdar://129223343

Reviewed by Andres Gonzalez.

A lot of memory is wasted in AX isolated tree nodes storing boolean
attributes in an expensive HashMap. By storing some of the most common
ones in a bitfield we can save memory. The most rare ones are still a
good fit for the HashMap.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::AXIsolatedObject):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::boolAttributeValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
(WebCore::AXIsolatedObject::setPropertyFlag):
(WebCore::AXIsolatedObject::hasPropertyFlag const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/279938@main
